eng/download-source-built-archive.sh

```diff
@@ -89,39 +94,77 @@ function DownloadArchive {
     versionDelimiter="-"
   fi
 
-  # Assets for the default RID are hosted in builds.dotnet.microsoft.com
-  if [[ $artifactsRid == $defaultArtifactsRid ]]; then
-    useCILocation=false
-  fi
-
   archiveVersion="${versionDelimiter}${archiveVersion}${versionDelimiter}"
 
-  baseUrl="https://builds.dotnet.microsoft.com/dotnet/source-build"
-  if [[ "$useCILocation" == true ]]; then
-    baseUrl="https://ci.dot.net/public/source-build"
+  # Build array of base URLs to try in order
+  local baseUrls=()
+  
+  # Assets for the default RID are hosted in builds.dotnet.microsoft.com first
+  if [[ $artifactsRid == $defaultArtifactsRid ]]; then
+    baseUrls+=("https://builds.dotnet.microsoft.com/dotnet/source-build")
+  fi
+  
+  # Always try public CI location
+  baseUrls+=("https://ci.dot.net/public/source-build")
+  
+  # Try internal CI location if storage key is provided
+  if [[ -n "$storageKey" ]]; then
+    baseUrls+=("https://ci.dot.net/internal/source-build")
```

Comment:
nit: You can avoid the string compare below to [ci.dot.net/internal](https://github.com/dotnet/dotnet/pull/3219/files#diff-45f7283695533c504fd2655f72aabfc6b924e17e4603028153d4aef5e5aa7f8aR139) by updating the baseUrls array to be a series of key and optional value pairs (e.g. "https://ci.dot.net/public/source-build" and "https://ci.dot.net/internal/source-build=$storageKey")